cmd/cstor-sidecar/ccrd/controller-cstor.go

```diff
@@ -0,0 +1,340 @@
+// processNextWorkItem function in order to read and process a message on the
+// workqueue.
+func (c *Controller) runWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem will read a single work item off the workqueue and
+// attempt to process it, by calling the syncHandler.
+func (c *Controller) processNextWorkItem() bool {
+	obj, shutdown := c.workqueue.Get()
+
+	if shutdown {
+		return false
+	}
+
+	// We wrap this block in a func so we can defer c.workqueue.Done.
+	err := func(obj interface{}) error {
+		// We call Done here so the workqueue knows we have finished
+		// processing this item. We also must remember to call Forget if we
+		// do not want this work item being re-queued. For example, we do
+		// not call Forget if a transient error occurs, instead the item is
+		// put back on the workqueue and attempted again after a back-off
+		// period.
+		defer c.workqueue.Done(obj)
+		var q QueueLoad
+		var ok bool
+		// We expect strings to come off the workqueue. These are of the
+		// form namespace/name. We do this as the delayed nature of the
+		// workqueue means the items in the informer cache may actually be
+		// more up to date that when the item was initially put onto the
+		// workqueue.
+		if q, ok = obj.(QueueLoad); !ok {
+			// As the item in the workqueue is actually invalid, we call
+			// Forget here else we'd go into a loop of attempting to
+			// process a work item that is invalid.
+			c.workqueue.Forget(obj)
+			runtime.HandleError(fmt.Errorf("expected string in workqueue but got %#v", obj))
+			return nil
+		}
+		// Run the syncHandler, passing it the namespace/name string of the
+		// spc resource to be synced.
+		if err := c.syncHandler(q.key, q.operation); err != nil {
+			return fmt.Errorf("error syncing '%s': %s", q.key, err.Error())
+		}
+		// Finally, if no error occurs we Forget this item so it does not
+		// get queued again until another change happens.
+		c.workqueue.Forget(obj)
+		glog.Infof("Successfully synced '%s'", q.key)
+		return nil
+	}(obj)
+
+	if err != nil {
+		runtime.HandleError(err)
+		return true
+	}
+
+	return true
+}
+
+// syncHandler compares the actual state with the desired, and attempts to
+// converge the two. It then updates the Status block of the spcUpdated resource
+// with the current status of the resource.
+func (c *Controller) syncHandler(key, operation string) error {
+	// Convert the namespace/name string into a distinct namespace and name
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("invalid resource key: %s", key))
+		return nil
+	}
+
+	// Get the spcUpdated resource with this namespace/name
+	if namespace == "" {
+		namespace = "default"
+	}
+
+	//spcUpdated, err := c.spcLister.StoragePoolClaims(namespace).Get(name)
+	_ = namespace
+	cstorCrdUpdated, err := c.clientset.OpenebsV1alpha1().CstorCrds(namespace).Get(name, metav1.GetOptions{})
```

Comment:
How can we be sure that this request needs to go to OpenebsV1alpha1()?
This needs a check before invoking a particular version?
The event might be for a different version?

Reply:
This depends on the autogen code we had given v1alpha1 before and the corresponding autogenerated has the name.

Reply:
We had decided to go with writing all the types onto the same types.go file of pkg/apis... and they will have same name